contracts/core/Allo.sol

```diff
@@ -1,79 +1,368 @@
         Metadata metadata;
-        bool active;
     }
 
-    /**
-     * @notice Initializes the contract after an upgrade
-     * @dev In future deploys of the implementation, an higher version should be passed to reinitializer
-     */
-    function initialize() public reinitializer(1) {}
+    /// @notice Fee denominator
+    uint256 public constant FEE_DENOMINATOR = 1e18;
+
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Fee percentage
+    /// @dev 1e18 = 100%, 1e17 = 10%, 1e16 = 1%, 1e15 = 0.1%
+    uint256 public feePercentage;
+
+    /// @notice Incremental index
+    uint256 private _poolIndex;
+
+    /// @notice Nonce for cloning strategies
+    uint256 private _nonce;
+
+    /// @notice Allo treasury
+    address payable public treasury;
+
+    /// @notice Registry of pool creators
+    Registry public registry;
+
+    /// @notice Pool.id -> Pool
+    mapping(uint256 => Pool) public pools;
+
+    /// @notice Strategy -> bool
+    mapping(address => bool) public approvedStrategies;
+
+    /// @notice Strategy -> bool
+    mapping(address => bool) public usedStrategies;
+
+    /// ======================
+    /// ======= Events =======
+    /// ======================
+
+    event PoolCreated(
+        uint256 indexed poolId,
+        bytes32 indexed identityId,
+        address allocationStrategy,
+        address distributionStrategy,
+        address token,
+        uint256 amount,
+        Metadata metadata
+    );
 
-    // Public getter on the pools mapping
-    /// @notice calls out to the registry to get the identity metadata
-    function getPoolInfo(uint256 _poolId) external view returns (PoolData memory, string memory) {
-        // Implement the function here
+    event PoolMetadataUpdated(uint256 indexed poolId, Metadata metadata);
+
+    event PoolFunded(uint256 indexed poolId, uint256 amount, uint256 fee);
+
+    event PoolClosed(uint256 indexed poolId);
+
+    event TreasuryUpdated(address treasury);
+
+    event FeeUpdated(uint256 fee);
+
+    event RegistryUpdated(address registry);
+
+    /// ====================================
+    /// =========== Intializer =============
+    /// ====================================
+
+    /// @notice Initializes the contract after an upgrade
+    /// @dev During upgrade -> an higher version should be passed to reinitializer
+    /// @param _registry The address of the registry
+    /// @param _treasury The address of the treasury
+    /// @param _feePercentage The fee percentage
+    function initialize(address _registry, address payable _treasury, uint256 _feePercentage) public reinitializer(1) {
+        _initializeOwner(msg.sender);
+
+        registry = Registry(_registry);
+        treasury = _treasury;
+        feePercentage = _feePercentage;
+
+        emit RegistryUpdated(_registry);
+        emit TreasuryUpdated(_treasury);
+        emit FeeUpdated(_feePercentage);
+
+        __Multicall_init();
```

Comment:
I need another bit of time to think through if there are any safety implications of multicall here. I think it's fine.

Reply:
Agreed this is alright over the call

Reply:
fixed via 327d43ceb2806c59aad6dead10a9d4f2affb961a